File: bnhsdk/ios/Classes/BinahPreviewFactory.swift
//
//  BinahPreviewFactory.swift
//  binah_flutter_sdk
//
//  Created by Giora Vered on 26/03/2023.
//

import Foundation


class BinahPreviewFactory: NSObject, FlutterPlatformViewFactory {
    
    static let shared: BinahPreviewFactory = BinahPreviewFactory()
    static let cameraPreviewId = "plugins.binah.ai/camera_preview_view"
    
    private var cameraPreview: BinahCameraPreview?
    private var imageDataSource: ImageDataSource?
    
    private override init() { }
    
    func create(withFrame frame: CGRect, viewIdentifier viewId: Int64, arguments args: Any?) -> FlutterPlatformView {
        let preview = BinahCameraPreview()
        preview.setDataSource(imageDataSource: imageDataSource)
        return preview
    }
    
    func setDataSource(imageDataSource: ImageDataSource) {
        self.imageDataSource = imageDataSource
        cameraPreview?.setDataSource(imageDataSource: imageDataSource)
    }
    
    
    
}

File: bnhsdk/android/src/main/kotlin/ai/binah/binah_flutter_sdk/BinahPreviewFactory.kt
package ai.binah.binah_flutter_sdk

import android.content.Context
import io.flutter.plugin.common.StandardMessageCodec
import io.flutter.plugin.platform.PlatformView
import io.flutter.plugin.platform.PlatformViewFactory

object BinahPreviewFactory: PlatformViewFactory(StandardMessageCodec.INSTANCE) {

    const val cameraPreviewId = "plugins.binah.ai/camera_preview_view"
    private var cameraPreview: BinahCameraPreview? = null
    private var imageDataSource: ImageDataSource? = null

    override fun create(context: Context, viewId: Int, args: Any?): PlatformView {
        return BinahCameraPreview(context).also {
            cameraPreview = it
            cameraPreview?.setDataSource(imageDataSource ?: return@also)
        }
    }

    fun setDataSource(imageDataSource: ImageDataSource) {
        this.imageDataSource = imageDataSource
        cameraPreview?.setDataSource(imageDataSource)
    }
}